src/main/java/picard/sam/SAMSplitter.java

```diff
@@ -0,0 +1,183 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam;
+
+import htsjdk.samtools.*;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.ProgressLogger;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineProgramProperties;
+import org.broadinstitute.barclay.help.DocumentedFeature;
+import picard.PicardException;
+import picard.cmdline.CommandLineProgram;
+import picard.cmdline.StandardOptionDefinitions;
+import picard.cmdline.programgroups.SamOrBam;
+
+import java.io.File;
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * <p/>
+ * Splits the input queryname sorted or querygrouped SAM/BAM file and writes it into
+ * multiple BAM files of approximately equal size. This will retain the sort order
+ * within each output BAM and if the BAMs are concatenated in order (output files are named
+ * numerically) the order of the reads will match the original BAM. As this tool is
+ * intended for unmapped input files, all secondary and supplementary reads will be removed.
+ */
+@CommandLineProgramProperties(
+        summary = SAMSplitter.USAGE_SUMMARY + SAMSplitter.USAGE_DETAILS,
+        oneLineSummary = SAMSplitter.USAGE_SUMMARY,
+        programGroup = SamOrBam.class)
+@DocumentedFeature
+public class SAMSplitter extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Splits a SAM or BAM file to multiple BAMs.";
+    static final String USAGE_DETAILS = "This tool splits the input SAM/BAM file into multiple BAM files " +
+            "This can be used to split a large unmapped BAM in order to parallelize alignment."+
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SAMSplitter \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT_DIR=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000" +
+            "     SPLIT_TO_N_TEMPLATES=24000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N", doc = "Split to have approximately N templates per output file.")
```

Comment:
I would argue that we should have `total_input_templates` not `total_input_reads`, since then we can just increment a "template" counter each time we see the primary alignment of either a fragment or the first end of a pair.  This keeps reads from the same template together, and also gets rid of the annoying `PAIRED_INPUT` argument (what if we have a mix).